Notify player when the research queue is empty

akien-mga opened the pull request:
Part of #869. What is missing is to add a check whether there are
actual skills left to be researched or not, and also make sure the
warning is displayed if there are no researches in queue from the start
(i.e. not only after a valid research is added).

akien-mga committed:
Notify player when the research queue is empty

Part of #869. What is missing is to add a check whether there are
actual skills left to be researched or not, and also make sure the
warning is displayed if there are no researches in queue from the start
(i.e. not only after a valid research is added).

akien-mga commented:
Actually rereading my changes, I see that mimicking the code for the treasury tile is not enough, as the treasury notification's "looping" come from being triggered repeatedly by workers, while the "no research in queue" message will only be triggered once after adding a completed research.

I'll see if I can make it appear again every 30s until a new research has been selected.

akien-mga commented:
> also make sure the warning is displayed if there are no researches in queue from the start
> (i.e. not only after a valid research is added).

Looks like the current code handles this case fine already thanks to the "null research entity" that gets added when no researches are selected.

akien-mga committed:
Loop the "no research in queue" warning

akien-mga commented:
The warning should now loop properly. The last thing needed to fix #869 properly would be to check if there are still skills that could be unlocked before printing the warning (i.e. don't print a warning if a seat has researches 100% of the tree). I wouldn't be against a small hint from @hwoarangmy for this one (not that I couldn't find it myself, but I'm on holidays with my other half and have little time to dig into code :p).

hwoarangmy commented:
> i.e. don't print a warning if a seat has researches 100% of the tree

You have to check if all researches are done, not allowed or depend on a research not allowed. To do that, you have to create a new function in the ResearchManager that takes a Seat as parameter and checks Research::canBeResearched to know if a research is possible.

akien-mga commented:
> You have to check if all researches are done, not allowed or depend on a research not allowed. To do that, you have to create a new function in the ResearchManager that takes a Seat as parameter and checks Research::canBeResearched to know if a research is possible.

Great, I'll have a look at that. That was more or less what I had hinted while looking the ResearchManager and Seat files, but I just wanted to check if you had something bigger in mind with regard to a potential "autofill" button that should generate a sensible queue for the player.

akien-mga commented:
> To do that, you have to create a new function in the ResearchManager that takes a Seat as parameter and checks Research::canBeResearched to know if a research is possible.

Looking at the current code, it feels like such a function would fit more next to the other research-related functions of game/SeatFunctions.cpp, don't you think? Especially since it could then access mResearchDone directly, and this removes the need to pass the seat as a parameter.

It might make sense to move most of the research-related functions in SeatFunctions.cpp to the ResearchManager in a second time though.

akien-mga commented:
Actually I'll try to work it out in the ResearchManager class and just expose the needed mResearchDone variable in the Seat.

hwoarangmy commented:
I don't have any strong opinion about what should be where. What I would request is to let all everything related to the GUI links (button name, path in research, path in GUI, points for research, ...) in the same place so that if someone wants to create a new spell, he doesn't have to check everywhere in the code where he should add something. And since researches are shared by rooms/traps/spells, it made sense to put the code there.
But if you see a better suiting place, please feel free to change

Bertram25 commented:
> related to the GUI links (button name, path in research, path in GUI, points for research, ...) in the same place so that if someone wants to create a new spell, he doesn't have to check everywhere in the code where he should add something. And since researches are shared by rooms/traps/spells, it made sense to put the code there.

+1 I guess that's the main architectural point here.

> But if you see a better suiting place, please feel free to change

In fact, if seomeone sees something, please discuss it first (as usual). [Yeah, I know, I'm an egomaniac dictator. ;)]

> Actually I'll try to work it out in the ResearchManager class and just expose the needed mResearchDone variable in the Seat.

All that said, putting the logic in the Research Manager and exposing the result in the Seat sounds pretty relevant. :)

Bertram25 commented on source/game/PlayerFunctions.cpp:
```diff
@@ -250,13 +251,27 @@ void Player::notifyTeamFighting(Player* player, Tile* tile)
+    {
+        mNoResearchInQueueTime = NOTIFICATION_TIME_COUNT;
+
+        std::string chatMsg = "Your research queue is empty, while there are still skills that could be unlocked.";
```
This reminds me, I'll rename all that to skill one day :P

akien-mga replied:
Agreed :)

Bertram25 commented:
Btw, it's cool to see you're opening the bowels of the beast. ;) @akien-mga 
So far, the code looks good, IMHO.

akien-mga commented:
I've done a first attempt at counting the number of remaining researches per seat, but I don't manage to call the function with the seat as argument, when in the seat source file.

Here is my diff: http://www.hastebin.com/xurokujire.coffee

And I get this error:

```
/home/akien/Projects/opendungeons/OpenDungeons-akien/source/game/SeatFunctions.cpp: In member function ‘void Seat::setNextResearch(ResearchType)’:
/home/akien/Projects/opendungeons/OpenDungeons-akien/source/game/SeatFunctions.cpp:1077:63: error: cannot call member function ‘std::vector<ResearchType> ResearchManager::getRemainingResearchesForSeat(const Seat*) const’ without object
         if(ResearchManager::getRemainingResearchesForSeat(this).size() == 0)
                                                               ^
```

Bertram25 commented:
This:
`if(ResearchManager::getRemainingResearchesForSeat(this).size() == 0)`
Such call (`MyNameSpace::MyFunction()`) is to be done when your function is static. Or if you prefer not linked to an object.
(Yes, classes can be seen as namespaces for static functions)

You'll have to add the `static` word in your function declaration, just like it is done with `listAllResearches()`

Hope this will help. :)

akien-mga commented:
Thanks, that seems to do the trick :-)

akien-mga committed:
Do not show empty queue notification when all skills have been researched

Also use separate timer lengths for the treasury and research notifications
as asked by @hwoarangmy.'

akien-mga commented:
Done, now my pull request should be ready :)
Please note that I haven't tested my last commit extensively yet. It does seem to work though from a first test with unlocking all researches using the console command.

akien-mga committed:
Use a boolean to know if the seat has unlocked all researches

akien-mga committed:
Do not loop the warning if all researches were unlocked

This could happen if a player gets the warning, and then unlocks all researches
with the console cheat (or maybe when unlocking the last research without cheat
during the timer).

akien-mga commented:
I've changed the function to return a boolean instead (the name is a bit wonky, but I couldn't come up with something better). I've also added a small change to make sure that the warning does not loop when the players unlock their last research during the timer.

akien-mga committed:
Improve the test based on @hwoarangmy's feedback

hwoarangmy commented:
Am I testing @akien-mga code ???

hwoarangmy commented:
First thing, the message triggers even if you don't own a library. That may be acceptable, though

hwoarangmy commented:
Works well. Mergeable for me

Bertram25 commented:
> First thing, the message triggers even if you don't own a library. That may be acceptable, though

I suggest we open an issue for that. But let's merge one of the most c++ implied PR from @akien-mga first. ;)
